File: packages/template-ts/src/parser/standalone.ts
// Standalone line groups (LEX-14, LEX-15).

export interface TagRange {
  // String indexes of the tag including any wrapper.
  start: number;
  end: number;
  echo: boolean;
}

export interface Range {
  start: number;
  end: number;
}

// Returns the string index ranges that standalone line groups remove from the output.
export function standaloneRanges(text: string, tags: readonly TagRange[]): Range[] {
  const lineStarts = [0];
  for (let i = 0; i < text.length; i++) {
    if (text.charCodeAt(i) === 0x0a) lineStarts.push(i + 1);
  }
  const lineCount = lineStarts.length;
  const lineOf = (index: number): number => {
    let low = 0;
    let high = lineCount - 1;
    while (low < high) {
      const mid = (low + high + 1) >> 1;
      if ((lineStarts[mid] as number) <= index) low = mid;
      else high = mid - 1;
    }
    return low;
  };
  const lineEnd = (line: number): number => (line + 1 < lineCount ? (lineStarts[line + 1] as number) : text.length);

  const tagsByLine: TagRange[][] = Array.from({ length: lineCount }, () => []);
  const sorted = [...tags].sort((a, b) => a.start - b.start);
  for (const tag of sorted) (tagsByLine[lineOf(tag.start)] as TagRange[]).push(tag);

  const ranges: Range[] = [];
  let line = 0;
  while (line < lineCount) {
    let groupEnd = line;
    let hasTag = false;
    let hasEcho = false;
    const groupTags: TagRange[] = [];
    for (let cursor = line; cursor <= groupEnd; cursor++) {
      for (const tag of tagsByLine[cursor] as TagRange[]) {
        hasTag = true;
        if (tag.echo) hasEcho = true;
        groupTags.push(tag);
        const endLine = lineOf(Math.max(tag.start, tag.end - 1));
        if (endLine > groupEnd) groupEnd = endLine;
      }
    }
    if (hasTag && !hasEcho) {
      const start = lineStarts[line] as number;
      const end = lineEnd(groupEnd);
      if (onlyWhitespaceOutside(text, start, end, groupTags)) ranges.push({ start, end });
    }
    line = groupEnd + 1;
  }
  return ranges;
}

function onlyWhitespaceOutside(text: string, start: number, end: number, tags: readonly TagRange[]): boolean {
  let index = start;
  for (const tag of tags) {
    if (!whitespaceOnly(text, index, tag.start)) return false;
    index = tag.end;
  }
  return whitespaceOnly(text, index, end);
}

function whitespaceOnly(text: string, start: number, end: number): boolean {
  for (let i = start; i < end; i++) {
    const code = text.charCodeAt(i);
    if (code !== 0x20 && code !== 0x09 && code !== 0x0d && code !== 0x0a) return false;
  }
  return true;
}

File: packages/template-go/parser/standalone.go
package parser

import "sort"

// tagRange is the byte range of a tag including its wrapper.
type tagRange struct {
	start, end int
	echo       bool
}

// byteRange is a removed range of the source.
type byteRange struct {
	start, end int
}

// standaloneRanges implements LEX-14: the byte ranges that standalone line groups remove.
func standaloneRanges(text string, tags []tagRange) []byteRange {
	lineStarts := []int{0}
	for i := 0; i < len(text); i++ {
		if text[i] == '\n' {
			lineStarts = append(lineStarts, i+1)
		}
	}
	lineCount := len(lineStarts)
	lineOf := func(index int) int {
		return sort.Search(lineCount, func(i int) bool { return lineStarts[i] > index }) - 1
	}
	lineEnd := func(line int) int {
		if line+1 < lineCount {
			return lineStarts[line+1]
		}
		return len(text)
	}
	sorted := append([]tagRange(nil), tags...)
	sort.Slice(sorted, func(i, j int) bool { return sorted[i].start < sorted[j].start })
	byLine := make([][]tagRange, lineCount)
	for _, tag := range sorted {
		line := lineOf(tag.start)
		byLine[line] = append(byLine[line], tag)
	}
	var ranges []byteRange
	for line := 0; line < lineCount; {
		groupEnd := line
		hasTag, hasEcho := false, false
		var groupTags []tagRange
		for cursor := line; cursor <= groupEnd; cursor++ {
			for _, tag := range byLine[cursor] {
				hasTag = true
				if tag.echo {
					hasEcho = true
				}
				groupTags = append(groupTags, tag)
				last := tag.end - 1
				if last < tag.start {
					last = tag.start
				}
				if endLine := lineOf(last); endLine > groupEnd {
					groupEnd = endLine
				}
			}
		}
		if hasTag && !hasEcho {
			start, end := lineStarts[line], lineEnd(groupEnd)
			if onlyWhitespaceOutside(text, start, end, groupTags) {
				ranges = append(ranges, byteRange{start, end})
			}
		}
		line = groupEnd + 1
	}
	return ranges
}

func onlyWhitespaceOutside(text string, start, end int, tags []tagRange) bool {
	index := start
	for _, tag := range tags {
		if !whitespaceOnly(text, index, tag.start) {
			return false
		}
		index = tag.end
	}
	return whitespaceOnly(text, index, end)
}

func whitespaceOnly(text string, start, end int) bool {
	for i := start; i < end; i++ {
		switch text[i] {
		case ' ', '\t', '\r', '\n':
		default:
			return false
		}
	}
	return true
}
